mlir/include/mlir/Dialect/MemRef/TransformOps/MemRefTransformOps.td

```diff
@@ -144,6 +144,71 @@ def ApplyResolveRankedShapedTypeResultDimsPatternsOp : Op<Transform_Dialect,
 }
 
 def Transform_MemRefAllocOp : Transform_ConcreteOpType<"memref.alloc">;
+def Transform_MemRefAllocaOp : Transform_ConcreteOpType<"memref.alloca">;
+
+def MemRefAllocaToGlobalOp :
+  Op<Transform_Dialect, "memref.alloca_to_global",
+     [TransformOpInterface,
+      DeclareOpInterfaceMethods<MemoryEffectsOpInterface>,
+      DeclareOpInterfaceMethods<TransformOpInterface>]> {
+  let description = [{
+    Inserts a new `memref.global` for each provided `memref.alloca` into the
+    provided module and replaces it with a `memref.get_global`. This is useful,
+    for example, for allocations that should reside in the shared memory of
+    a GPU, which have to be declared as globals.
+
+    #### Example
+
+    Consider the following transform op:
+
+    ```mlir
+    %get_global, %global =
+        transform.memref.alloca_to_global %alloca in %module
+          : (!transform.op<"builtin.module">, !transform.op<"memref.alloca">)
+            -> (!transform.any_op, !transform.any_op)
+    ```
+
+    and the following input payload:
+
+    ```mlir
+    module {
+      func.func @func() {
+        %alloca = memref.alloca() : memref<2x32xf32>
+        // usages of %alloca...
+      }
+    }
+    ```
+
+    then applying the transform op to the payload would result in the following
+    output IR:
+
+    ```mlir
+    module {
+      memref.global "private" @alloc : memref<2x32xf32>
+      func.func @func() {
+        %alloca = memref.get_global @alloc : memref<2x32xf32>
+        // usages of %alloca...
+      }
+    }
+    ```
+
+    #### Return modes
+
+    Emits a definite failure if not exactly one `module` payload op was provided
+    or any of the `alloca` payload ops is not inside that module, and succeeds
+    otherwise. The returned handles refer to the `memref.get_global` and
+    `memref.global` ops that were inserted by the transformation.
+  }];
+
+  let arguments = (ins Transform_ConcreteOpType<"builtin.module">:$module,
```

Comment:
I don't think we want to limit this argument to be specifically `builtin.module`. In the use case the documentation describes, this can clearly also be a `gpu.module`. But in general it is any symbol table.

Reply:
OK, good point.

Coming back to the question of whether we need this argument at all: `memref.get_global` expects the symbol to be in the nearest `SymbolTable` (see [verifier](https://github.com/llvm/llvm-project/blob/13e9a568/mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp#L1635-L1639)). This means that, given the `memref.alloca` payload, there is only one symbol table to insert the symbol into. Do we really need to pass in that op explicitly, then?

Reply:
Allowing any op with the `SymbolTable` trait in the newest revision.

Reply:
>  Do we really need to pass in that op explicitly, then?

I can imagine a hypothetical future where we can refer to globals through nested named scopes. But it's not the case today, so both ways are fine.